crates/evm/src/call.rs

```diff
@@ -28,14 +29,115 @@ pub struct CallMessage {
 }
 
 impl<C: sov_modules_api::Context> Evm<C> {
+    /// Executes system events for the current block and push tx to pending_transactions.
+    ///
+    /// Returns cumulative used gas.
+    pub(crate) fn execute_system_events(
+        &self,
+        system_events: Vec<SystemEvent>,
+        working_set: &mut WorkingSet<C>,
+    ) {
+        if system_events.is_empty() {
+            return;
+        }
+
+        let block_env = self
+            .block_env
+            .get(working_set)
+            .expect("Pending block must be set");
+
+        let cfg = self.cfg.get(working_set).expect("Evm config must be set");
+        let cfg_env: CfgEnvWithHandlerCfg = get_cfg_env(&block_env, cfg, None);
+
+        let l1_fee_rate = self
+            .l1_fee_rate
+            .get(working_set)
+            .expect("L1 fee rate must be set");
+
+        let l1_block_hash_exists = self
+            .accounts
+            .get(&L1BlockHashList::address(), working_set)
+            .is_some();
+        if !l1_block_hash_exists {
+            tracing::error!("System contract not found: L1BlockHashList");
+            return;
+        }
+
+        let system_nonce = self
+            .accounts
+            .get(&SYSTEM_SIGNER, working_set)
+            .map(|acc| acc.info.nonce)
+            .unwrap_or(0);
+
+        let db: EvmDb<'_, C> = self.get_db(working_set);
+        let system_txs = create_system_transactions(system_events, system_nonce, cfg_env.chain_id);
+
+        let mut citrea_handler_ext = CitreaExternal::new(l1_fee_rate);
+        let block_number = block_env.number;
+        let tx_results = executor::execute_system_txs(
+            db,
+            block_env,
+            &system_txs,
+            cfg_env,
+            &mut citrea_handler_ext,
+        );
+
+        let mut cumulative_gas_used = 0;
+        let mut log_index_start = 0;
+        if let Some(tx) = self.pending_transactions.last(working_set) {
+            cumulative_gas_used = tx.receipt.receipt.cumulative_gas_used;
+            log_index_start = tx.receipt.log_index_start + tx.receipt.receipt.logs.len() as u64;
+        }
```

Comment:
why do we check this?

Reply:
let's double down here, 

if pending_transactions is not empty, panic with system transactions should be the first tx in block.

cum_gas_used and log_index_start = 0 always in this func.

Reply:
I can init them with 0 but I am not quite sure about our l1 fee strategy, if we end up with exec a sys tx in the end of the block then 0 will be an incorrect init value